Add easier to read print option

YigitElma opened the pull request:
Add an option to print start and end errors next to each other for easier comparison.
Resolves #1112 

```
==============================================================================================================
                                                       Start  -->   End
Total (sum of squares)                             6.877e-12  -->   6.877e-12, 
Maximum absolute Force error:                      7.852e+00  -->   7.852e+00 (N)
Minimum absolute Force error:                      1.009e-05  -->   1.009e-05 (N)
Average absolute Force error:                      1.148e-01  -->   1.148e-01 (N)
Maximum absolute Force error:                      2.546e-06  -->   2.546e-06 (normalized)
Minimum absolute Force error:                      3.274e-12  -->   3.274e-12 (normalized)
Average absolute Force error:                      3.722e-08  -->   3.722e-08 (normalized)
R boundary error:                                  0.000e+00  -->   0.000e+00 (m)
Z boundary error:                                  0.000e+00  -->   0.000e+00 (m)
Fixed Psi error:                                   0.000e+00  -->   0.000e+00 (Wb)
Fixed pressure profile error:                      0.000e+00  -->   0.000e+00 (Pa)
Fixed iota profile error:                          0.000e+00  -->   0.000e+00 (dimensionless)
Fixed sheet current error:                         0.000e+00  -->   0.000e+00 (~)
==============================================================================================================
```

- [x] Add a test that checks length of `print_value_fmt` of each objective and ensures that they are shorter than `print_result_width` value

YigitElma commented on desc/optimize/optimizer.py:
```diff
@@ -349,29 +349,45 @@ def optimize(  # noqa: C901 - FIXME: simplify this
+            state_0 = [things0[things.index(t)] for t in objective.things]
+            state = [things[things.index(t)] for t in objective.things]
+
+            print_pretty = (
```
Tbh, I don't like this. But you guys asked for the old print to remain. I guess we can get rid of this and always print the comparison version

dpanici replied:
We meant, the old use of `print_value` should be the same, if I were to do `obj.print_value` with only one set of args. 

I think we all want the optimizer to by default print the comparisons in the way you've changed them to, so you don't need this conditional here

dpanici commented on desc/objectives/objective_funs.py:
```diff
@@ -1072,24 +1090,37 @@ def jvp_unscaled(self, v, x, constants=None):
         return self._jvp(v, x, constants, "compute_unscaled")
 
-    def print_value(self, *args, **kwargs):
+    def print_value(self, val, val0=None, **kwargs):
```
I'd prefer something like `args, args0` instead of `val`, as `val` makes it sound like you are printing that value, when really these are the arguments to the objective compute function

dpanici commented on desc/objectives/objective_funs.py:
```diff
@@ -1106,42 +1137,67 @@ def print_value(self, *args, **kwargs):
             fmin = jnp.min(f)
             fmean = jnp.mean(f * w) / jnp.mean(w)
 
+            f0 = jnp.abs(f0) if abserr else f0 if val0 is not None else f
```
would prefer to bite the bullet and split this up across multiple lines as at least to me it is not super legible. I know what it is doing but the double if on the same line is hard to follow

dpanici commented on desc/objectives/objective_funs.py:
```diff
@@ -1106,42 +1137,67 @@ def print_value(self, *args, **kwargs):
                 "Maximum "
                 + ("absolute " if abserr else "")
-                + self._print_value_fmt.format(fmax)
+                + print_value_fmt.format(f0max, fmax)
```
If you provide more values to .format then there are spaces in the string for such values, it is fine with that? No errors/warnings are thrown?

when `val0` is None there is only a single slot for a value in the string

dpanici replied:
hm seems so, ok nvm, maybe though a comment here saying "only first value will be printed which is fmin when `val0` is None" would make it clearer

dpanici reviewed (changes requested):
- make optimizer by default use the comparisons print that you implemented, we don't need the old logic in there for the old way we used to print in the optimizer.
- add a test for the comparison printing
- rerun any notebooks that do `.solve` or `.optimize` so that they output with the new comparison printing

YigitElma commented:
Are black and linting tests failing due to the high number of lines changed or some server error? I have tried to re-run them several times, but they fail anyway.

f0uriest commented:
> Are black and linting tests failing due to the high number of lines changed or some server error? I have tried to re-run them several times, but they fail anyway.

There was an outage earlier that killed a bunch of CI jobs, might still be kind of buggy. Usually just wait a few hours and retry

YigitElma commented:
Black and linting tests give a timeout type error. I assume that is because of the high number of lines changed (I decreased the resolution of 3D plot while I was re-running the notebooks, it still looks good, but it decreased the file size 30Mb!). Actual code changes are around 200 lines at most. The recommit passed everything locally

unalmis reviewed (changes requested):
I'd like all the -> to align with each other, so that printing looks like:
```
Start of solver                                     -->   End of solver
Total (sum of squares):                 6.409e+01   -->   1.523e-03 
Maximum absolute Quasi-symmetry error:  4.730e+02   -->   2.086e+00(T^4/m^2)
Minimum absolute Quasi-symmetry error:  4.422e-04   -->   2.614e-05(T^4/m^2)
Average absolute Quasi-symmetry error:  1.117e+01   -->   1.081e-01(T^4/m^2)
Maximum absolute Quasi-symmetry error:  4.183e+00   -->   1.845e-02(normalized)
Minimum absolute Quasi-symmetry error:  3.910e-06   -->   2.312e-07(normalized)
Average absolute Quasi-symmetry error:  9.879e-02   -->   9.560e-04(normalized)
Maximum absolute Force error:           1.516e+05   -->   1.611e+04(N)
Minimum absolute Force error:           2.431e+00   -->   2.379e-01(N)
Average absolute Force error:           5.922e+03   -->   7.447e+02(N)
Maximum absolute Force error:           8.198e-03   -->   8.715e-04(normalized)
Minimum absolute Force error:           1.315e-07   -->   1.287e-08(normalized)
Average absolute Force error:           3.203e-04   -->   4.028e-05(normalized)
R boundary error:                       0.000e+00   -->   0.000e+00(m)
Z boundary error:                       0.000e+00   -->   0.000e+00(m)
Fixed pressure profile error:           0.000e+00   -->   1.608e-13(Pa)
Fixed iota profile error:               0.000e+00   -->   0.000e+00(dimensionless)
Fixed Psi error:                        0.000e+00   -->   0.000e+00(Wb)
```

Here is the code I'd recommend using
```python
padding = 65   # make this large enough
thing_to_print = force_error
print(f"{'text to print':<{padding}}", thing_to_print)
```

dpanici commented on desc/objectives/_free_boundary.py:
```diff
@@ -265,35 +266,44 @@ def print_value(self, *args, **kwargs):
         abserr = jnp.all(self.target == 0)
 
-        def _print(fmt, fmax, fmin, fmean, norm, units):
+        def _print(fmt, fmax, fmin, fmean, f0max, f0min, f0mean, norm, unit):
```
why change `units` to `unit`? seems unnecessary

YigitElma replied:
Sorry my english is bad :laughing:

dpanici reviewed (changes requested):
`Maximum Rotational transform:  2.477e-01   -->   1.102e+00(dimensionless)`

I think we need another space btwn the number and the units, so it shouuld look like
`Maximum Rotational transform:  2.477e-01   -->   1.102e+00 (dimensionless)`

YigitElma commented:
> I'd like all the -> to align with each other, so that printing looks like:


So, I thought about it after @ddudt suggested yesterday, but I have couple concern. Let's say we specify a max length according to the longest objective print name, then even if we don't have that objective, we will have a spacing. For example in the text you sent
```
Maximum absolute Force error:           1.516e+05   -->   1.611e+04(N)
Minimum absolute Force error:           2.431e+00   -->   2.379e-01(N)
Average absolute Force error:           5.922e+03   -->   7.447e+02(N)
Maximum absolute Force error:           8.198e-03   -->   8.715e-04(normalized)
Minimum absolute Force error:           1.315e-07   -->   1.287e-08(normalized)
Average absolute Force error:           3.203e-04   -->   4.028e-05(normalized)
R boundary error:                       0.000e+00   -->   0.000e+00(m)
Z boundary error:                       0.000e+00   -->   0.000e+00(m)
Fixed pressure profile error:           0.000e+00   -->   1.608e-13(Pa)
Fixed iota profile error:               0.000e+00   -->   0.000e+00(dimensionless)
Fixed Psi error:                        0.000e+00   -->   0.000e+00(Wb)
```
I can add a for loop before the main print for loop to get the longest name and decide the padding according to that. But I would rather do that if everybody wants this.

f0uriest commented on docs/notebooks/tutorials/free_boundary_equilibrium.ipynb:
Check the printing for the boundary error objectives, they seem to be printing the same value for before and after

YigitElma replied:
Good catch, thanks!

f0uriest commented on docs/notebooks/tutorials/input.HELIOTRON_desc:
This shouldn't be committed, its generated by running one of the notebooks

f0uriest reviewed (dismissed):
I actually prefer to keep the different objectives with different line lengths, I find it helps me distinguish them visually, but if people feel strongly feel free to change them to make them all line up

Also I Updated requirements and precommit to format notebooks with black, I'd been doing it manually but this should prevent future annoyance.

YigitElma commented:
```
----------------------------------------------------------------------------------------------------
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
----------------------------------------------------------------------------------------------------

                                                       Start  -->   End
Total (sum of squares)                             6.877e-12  -->   6.877e-12, 
Maximum absolute Force error:                      7.852e+00  -->   7.852e+00 (N)
Minimum absolute Force error:                      1.009e-05  -->   1.009e-05 (N)
Average absolute Force error:                      1.148e-01  -->   1.148e-01 (N)
Maximum absolute Force error:                      2.546e-06  -->   2.546e-06 (normalized)
Minimum absolute Force error:                      3.274e-12  -->   3.274e-12 (normalized)
Average absolute Force error:                      3.722e-08  -->   3.722e-08 (normalized)
R boundary error:                                  0.000e+00  -->   0.000e+00 (m)
Z boundary error:                                  0.000e+00  -->   0.000e+00 (m)
Fixed Psi error:                                   0.000e+00  -->   0.000e+00 (Wb)
Fixed pressure profile error:                      0.000e+00  -->   0.000e+00 (Pa)
Fixed iota profile error:                          0.000e+00  -->   0.000e+00 (dimensionless)
Fixed sheet current error:                         0.000e+00  -->   0.000e+00 (~)

----------------------------------------------------------------------------------------------------
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
----------------------------------------------------------------------------------------------------
```

f0uriest commented on desc/optimize/optimizer.py:
```diff
@@ -296,9 +298,12 @@ def optimize(  # noqa: C901 - FIXME: simplify this
 
         if verbose > 0:
-            print("Starting optimization")
+            print("\nStarting optimization")
```
why do we need the extra newline here?

YigitElma replied:
We print many things, I felt like dividing the initialization and optimization part would be more organized. Using the divider like the `--------`s is a bit too much but I guess it makes it cleaner

YigitElma replied:
Don't you think this easier to focus?
```
Building objective: Bootstrap current self-consistency (Redl)
Precomputing transforms
Timer: Precomputing transforms = 832 ms
Timer: Objective build = 2.55 sec
Building objective: lcfs R
Building objective: lcfs Z
Building objective: fixed Psi
Building objective: fixed current
Building objective: fixed electron density
Building objective: fixed electron temperature
Building objective: fixed ion temperature
Building objective: fixed atomic number
Building objective: fixed sheet current
Building objective: self_consistency R
Building objective: self_consistency Z
Building objective: lambda gauge
Building objective: axis R self consistency
Building objective: axis Z self consistency
Timer: Objective build = 1.50 sec
Timer: Linear constraint projection build = 3.07 sec
Number of parameters: 37
Number of objectives: 2
Timer: Initializing the optimization = 7.39 sec

Starting optimization
Using method: lsq-exact
   Iteration     Total nfev        Cost      Cost reduction    Step norm     Optimality   
       0              1          7.871e-13                                    2.292e-13   
`gtol` condition satisfied.
         Current function value: 7.871e-13
         Total delta_x: 0.000e+00
         Iterations: 0
         Function evaluations: 1
         Jacobian evaluations: 1
Timer: Solution time = 9.66 sec
Timer: Avg time per step = 9.66 sec

--------------------------------------------------------------------------------------------------------------
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
--------------------------------------------------------------------------------------------------------------

                                                                 Start  -->   End
Total (sum of squares):                                      7.871e-13  -->   7.871e-13, 
Maximum absolute Bootstrap current self-consistency error:   1.013e-01  -->   1.013e-01 (T A m^-2)
Minimum absolute Bootstrap current self-consistency error:   1.010e-01  -->   1.010e-01 (T A m^-2)
Average absolute Bootstrap current self-consistency error:   1.012e-01  -->   1.012e-01 (T A m^-2)
Maximum absolute Bootstrap current self-consistency error:   1.256e-06  -->   1.256e-06 (normalized)
Minimum absolute Bootstrap current self-consistency error:   1.253e-06  -->   1.253e-06 (normalized)
Average absolute Bootstrap current self-consistency error:   1.255e-06  -->   1.255e-06 (normalized)
R boundary error:                                            0.000e+00  -->   0.000e+00 (m)
Z boundary error:                                            0.000e+00  -->   0.000e+00 (m)
Fixed Psi error:                                             0.000e+00  -->   0.000e+00 (Wb)
Fixed current profile error:                                 0.000e+00  -->   0.000e+00 (A)
Fixed electron density profile error:                        0.000e+00  -->   0.000e+00 (m^-3)
Fixed electron temperature profile error:                    0.000e+00  -->   0.000e+00 (eV)
Fixed ion temperature profile error:                         0.000e+00  -->   0.000e+00 (eV)
Fixed atomic number profile error:                           0.000e+00  -->   0.000e+00 (dimensionless)
Fixed sheet current error:                                   0.000e+00  -->   0.000e+00 (~)

--------------------------------------------------------------------------------------------------------------
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
--------------------------------------------------------------------------------------------------------------
```

f0uriest commented on desc/optimize/optimizer.py:
```diff
@@ -352,7 +357,13 @@ def optimize(  # noqa: C901 - FIXME: simplify this
-            print("Start of solver   -->   End of solver")
+            # put a divider
+            w_divider = 50
+            print("\n{:-<{}}".format("", print_result_width + w_divider))
```
I don't think we need 3 lines for a divider, 1 would be fine

f0uriest commented on desc/utils.py:
```diff
@@ -682,3 +682,6 @@ def broadcast_tree(tree_in, tree_out, dtype=int):
         raise ValueError("trees must be nested lists of dicts")
+
+
+print_result_width = 60
```
global constants like this should be in ALL_CAPS

f0uriest commented on tests/test_objective_funs.py:
```diff
@@ -2630,3 +2654,33 @@ def test_asymmetric_normalization():
+
+def test_objective_print_widths():
+    """Test that the objective's name is shorter than max."""
+    from desc.objectives.objective_funs import _Objective
```
this can be imported at the top

ddudt commented on desc/utils.py:
```diff
@@ -682,3 +682,6 @@ def broadcast_tree(tree_in, tree_out, dtype=int):
         raise ValueError("trees must be nested lists of dicts")
+
+
+PRINT_WIDTH = 60
```
Is this width is based on the bootstrap objective? We should include a comment in the code saying why we chose this hard-coded number

YigitElma replied:
Yes, current longest name is Bootstrap one. Actual value is 58 I guess but I like round numbers :D

YigitElma replied:
Yes, current longest name is Bootstrap one. Actual value is 58 I guess but I like round numbers :D

YigitElma replied:
Yes, current longest name is Bootstrap one. Actual value is 58 I guess but I like round numbers :D

YigitElma replied:
Yes, current longest name is Bootstrap one. Actual value is 58 I guess but I like round numbers :D

ddudt commented on desc/utils.py:
Would this be better to put in `desc/objectives/utils.py`?

YigitElma replied:
It was also called by `desc/optimize/optimizer.py`, that's why I didn't have strong preference.

ddudt reviewed (commented):
This looks so much better! 

I think the line breaks `\n` are excessive since we already separate the sections with the diving line `====`, but I'm fine with it if everyone else prefers having the line breaks

unalmis commented:
if you want a nitpick, my personal preference is one line of dashes "-" to a line of "= and/or +". either way thanks for doing this

YigitElma commented:
> if you want a nitpick, my personal preference is one line of dashes "-" to a line of "= and/or +". either way thanks for doing this

I tried `-` first, it looked a bit weak. At least for me, `=` acts better up to its name as divider :smile:

dpanici reviewed (changes requested):
I am going to be annoying but could we have a line break btwn objectives so you can tell them apart? otherwise it is just like 100 numbers and they are far from the objective names that say what each are...

dpanici reviewed (dismissed):
Nevermind, the units help distinguish

dpanici commented:
run `black` and `flake8` on repo then we will commit
